Mathlib/NumberTheory/AbelSummation.lean

```diff
@@ -129,45 +159,217 @@ theorem _root_.sum_mul_eq_sub_sub_integral_mul (ha : 0 ≤ a) (hab : a ≤ b)
+      f m * (∑ k ∈ Icc 0 m, c k) -
+        ∫ t in Set.Ioc (1 : ℝ) m, deriv f t * ∑ k ∈ Icc 0 ⌊t⌋₊, c k := by
+  convert sum_mul_eq_sub_integral_mul₀ c hc m hf_diff hf_int
+  all_goals rw [Nat.floor_natCast]
+
+end specialversions
+
+section limit
+
+open Filter Topology abelSummationProof intervalIntegral
+
+private theorem locallyintegrablemulsum (ha : 0 ≤ a) {g : ℝ → 𝕜} (hg : IntegrableOn g (Set.Ici a)) :
+    LocallyIntegrableOn (fun t ↦ g t * ∑ k ∈ Icc 0 ⌊t⌋₊, c k) (Set.Ici a) := by
+  refine (locallyIntegrableOn_iff isLocallyClosed_Ici).mpr fun K hK₁ hK₂ ↦ ?_
+  by_cases hK₃ : K.Nonempty
+  · have h_inf : a ≤ sInf K := (hK₁ (hK₂.sInf_mem hK₃))
+    refine IntegrableOn.mono_set ?_ (Bornology.IsBounded.subset_Icc_sInf_sSup hK₂.isBounded)
+    refine integrablemulsum _ (ha.trans h_inf) (hg.mono_set ?_)
+    exact (Set.Icc_subset_Ici_iff (Real.sInf_le_sSup _ hK₂.bddBelow hK₂.bddAbove)).mpr h_inf
+  · rw [Set.not_nonempty_iff_eq_empty.mp hK₃]
+    exact integrableOn_empty
+
+theorem tendsto_sum_mul_atTop_eq_sub_integral (hf_diff : ∀ t ∈ Set.Ici 0, DifferentiableAt ℝ f t)
+    (hf_int : IntegrableOn (deriv f) (Set.Ici 0)) {l : 𝕜}
+    (h_lim : Tendsto (fun n : ℕ ↦ f n * ∑ k ∈ Icc 0 n, c k) atTop (𝓝 l))
+    {g : ℝ → 𝕜} (hg_dom : (fun t ↦ deriv f t * ∑ k ∈ Icc 0 ⌊t⌋₊, c k) =O[atTop] g)
+    (hg_int : IntegrableAtFilter g atTop) :
+    Tendsto (fun n : ℕ ↦ ∑ k ∈ Icc 0 n, f k * c k) atTop
+      (𝓝 (l - ∫ t in Set.Ioi 0, deriv f t * ∑ k ∈ Icc 0 ⌊t⌋₊, c k)) := by
+  have h_lim' : Tendsto (fun n : ℕ ↦ ∫ t in Set.Ioc (0 : ℝ) n, deriv f t * ∑ k ∈ Icc 0 ⌊t⌋₊, c k)
+      atTop (𝓝 (∫ t in Set.Ioi 0, deriv f t * ∑ k ∈ Icc 0 ⌊t⌋₊, c k)) := by
+    refine Tendsto.congr (fun _ ↦ by rw [← integral_of_le (Nat.cast_nonneg _)]) ?_
+    refine intervalIntegral_tendsto_integral_Ioi _ ?_ tendsto_natCast_atTop_atTop
+    exact integrableOn_Ici_iff_integrableOn_Ioi.mp
+      <| (locallyintegrablemulsum c le_rfl hf_int).integrableOn_of_isBigO_atTop hg_dom hg_int
+  refine (h_lim.sub h_lim').congr (fun _ ↦ ?_)
+  rw [sum_mul_eq_sub_integral_mul' _ _ (fun t ht ↦ hf_diff _ ht.1)
+    (hf_int.mono_set Set.Icc_subset_Ici_self)]
+
+theorem tendsto_sum_mul_atTop_eq_sub_integral₀ (hc : c 0 = 0)
+    (hf_diff : ∀ t ∈ Set.Ici 1, DifferentiableAt ℝ f t)
+    (hf_int : IntegrableOn (deriv f) (Set.Ici 1)) {l : 𝕜}
+    (h_lim: Tendsto (fun n : ℕ ↦ f n * ∑ k ∈ Icc 0 n, c k) atTop (𝓝 l))
+    {g : ℝ → ℝ} (hg_dom : (fun t ↦ deriv f t * ∑ k ∈ Icc 0 ⌊t⌋₊, c k) =O[atTop] g)
+    (hg_int : IntegrableAtFilter g atTop) :
+    Tendsto (fun n : ℕ ↦ ∑ k ∈ Icc 0 n, f k * c k) atTop
+      (𝓝 (l - ∫ t in Set.Ioi 1, deriv f t * ∑ k ∈ Icc 0 ⌊t⌋₊, c k)) := by
+  have h_lim' : Tendsto (fun n : ℕ ↦ ∫ t in Set.Ioc (1 : ℝ) n, deriv f t * ∑ k ∈ Icc 0 ⌊t⌋₊, c k)
+      atTop (𝓝 (∫ t in Set.Ioi 1, deriv f t * ∑ k ∈ Icc 0 ⌊t⌋₊, c k)) := by
+    refine Tendsto.congr' (by
+      filter_upwards [eventually_ge_atTop 1] with _ h
+      rw [← integral_of_le (Nat.one_le_cast.mpr h)]) ?_
```

Comment:
Here it is perhaps better to write
```suggestion
    refine Tendsto.congr' ?_ ?_
    · filter_upwards [eventually_ge_atTop 1] with _ h
      rw [← integral_of_le (Nat.one_le_cast.mpr h)]
```
The inlined tactic proof is a bit hard to parse.

Reply:
I added another `have` before so that things get easier to parse hopefully.